indico/web/client/js/react/components/principals/ACLField.jsx

```diff
@@ -0,0 +1,186 @@
+// This file is part of Indico.
+// Copyright (C) 2002 - 2019 CERN
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import _ from 'lodash';
+import React, {useState, useEffect} from 'react';
+import PropTypes from 'prop-types';
+import {Button, List} from 'semantic-ui-react';
+import {Translate} from 'indico/react/i18n';
+import {UserSearch, GroupSearch} from '../principals/Search';
+import {useFetchPrincipals, usePermissionInfo} from '../principals/hooks';
+import {PendingPrincipalListItem, PrincipalListItem} from '../principals/items';
+import {getPrincipalList} from '../principals/util';
+
+import '../principals/PrincipalListField.module.scss';
+import PrincipalPermissions from './PrincipalPermissions';
+
+
+function buildACL(aclMap) {
+    return Object.entries(aclMap).map(([ident, permissions]) => (
+        [ident, _.sortBy(permissions.map(
+            p => ((p === 'fullAccess' || p === 'readAccess') ? '_' : '') + _.snakeCase(p)
+        ))]
+    ));
+}
+
+const isGroup = identifier => identifier.startsWith('Group:');
+
+/**
+ * The ACLField is a PrincipalField on steroids. In addition to the functionality
+ * present in ACLField, it keeps track of user permissions.
+ */
+const ACLField = (props) => {
+    const {
+        value, disabled, readOnly, onChange, onFocus, onBlur, withGroups,
+        favoriteUsersController, readAccessAllowed
+    } = props;
+    const [favoriteUsers, [handleAddFavorite, handleDelFavorite]] = favoriteUsersController;
+
+    const valueIds = value.map(([identifier]) => identifier);
+
+    // keep track of (user) details for each entry
+    const [identifierMap, setIdentifierMap] = useState({});
+
+    // Permission info - description of permission hierarchy
+    const [permissionManager, permissionInfo] = usePermissionInfo();
+
+    // fetch missing principals
+    const missingData = _.difference(valueIds, Object.keys(identifierMap));
+    useFetchPrincipals(missingData, setIdentifierMap);
+
+    // keep track of 'touched' state
+    const [touched, markTouched] = useState(false);
+
+    // keep track of permissions for each entry (ACL)
+    const acl = Object.assign({}, ...value.map(([identifier, permissions]) => (
+        {[identifier]: permissions.map(p => _.camelCase(p))}
```

Comment:
`{[identifier]: permissions.map(_.camelCase)}`?

Reply:
that will pass other arguments as well